### PR TITLE
[HttpClient][Mime] Add correct IDN flags for IDNA2008 compliance

### LINKED ISSUE #44091: Provide the right flags to IDN methods for HttpClient
### Symfony version(s) affected

5.3

### Description

Symfony HttpClient supports IDN but does not provide the right flags to the intl IDN methods (`idn_to_ascii` and `idn_to_utf8`) to be fully IDNA 2008 compliant (RFC 5891).
For example, `fußball.test` should be converted as `xn--fuball-cta.test` instead of `fussball.test`.

See [the affected line](https://github.com/symfony/symfony/blob/dd8c1dd9bac12563dcd962c9fd7e7ad1af9120b5/src/Symfony/Component/HttpClient/HttpClientTrait.php#L499) for conversion to ASCII.

### How to reproduce

```        
$client = HttpClient::create();
$response = $client->request('GET', $url);
```

If `$url` is `fußball.test` it should make a query to `xn--fuball-cta.test` instead of `fussball.test`.

### Possible Solution

The following flags should be provided to be the most compliant possible with IDNA 2008:
`IDNA_DEFAULT | IDNA_USE_STD3_RULES | IDNA_CHECK_BIDI | IDNA_CHECK_CONTEXTJ | IDNA_NONTRANSITIONAL_TO_ASCII`

### Additional Context

[Unicode IDNA compatibility processing](https://www.unicode.org/reports/tr46/)

### LINKED ISSUE #44092: Provide the right flags to IDN methods for email addresses conversion
### Symfony version(s) affected
5.3

### Description

Symfony Mailer converts the email addresses domain to A-label in order to be compliant with servers that does not support the SMTPUTF8 flag but this conversion is not fully IDNA2008 compliant.

See #44091 that is the equivalent report for HttpClient.

[This line](https://github.com/symfony/symfony/blob/dd8c1dd9bac12563dcd962c9fd7e7ad1af9120b5/src/Symfony/Component/Mime/Encoder/IdnAddressEncoder.php#L38) should be fixed.

### How to reproduce

```
$stream = new SocketStream();
$stream
    ->setHost($host)
    ->setPort($port)
    ->disableTls();
$transport = new SmtpTransport($stream);
$mailer = new Mailer($transport);
$email = (new Email())
    ->from($from)
    ->to($to)
    ->subject($subject)
    ->text($body);
$mailer->send($email);
```
If `$to` domain is fußball.test it should be converted to xn--fuball-cta.test instead of fussball.test.

### Possible Solution

The following flags should be provided to be the most compliant possible with IDNA 2008:
`IDNA_DEFAULT | IDNA_USE_STD3_RULES | IDNA_CHECK_BIDI | IDNA_CHECK_CONTEXTJ | IDNA_NONTRANSITIONAL_TO_ASCII`

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by j-bernard**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | #44091 #44092
| License       | MIT

Fix #44091
Fix #44092

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Can you please fix #44092 in the same PR since that's the same topic?

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 1]
```diff
@@ -456,7 +456,7 @@ private static function parseUrl(string $url, array $query = [], array $allowedS
                 throw new InvalidArgumentException(sprintf('Unsupported IDN "%s", try enabling the "intl" PHP extension or running "composer require symfony/polyfill-intl-idn".', $host));
             }
 
-            $host = \defined('INTL_IDNA_VARIANT_UTS46') ? idn_to_ascii($host, \IDNA_DEFAULT, \INTL_IDNA_VARIANT_UTS46) ?: strtolower($host) : strtolower($host);
+            $host = \defined('INTL_IDNA_VARIANT_UTS46') ? idn_to_ascii($host,  \IDNA_DEFAULT | \IDNA_USE_STD3_RULES | \IDNA_CHECK_BIDI | \IDNA_CHECK_CONTEXTJ | \IDNA_NONTRANSITIONAL_TO_ASCII, \INTL_IDNA_VARIANT_UTS46) ?: strtolower($host) : strtolower($host);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: first added space should be removed

**Comment by j-bernard** [comment]
I made the 2 updates. While making those, I noticed that in version 5.1 you allowed non-ASCII local-part without implementing [RFC6531](https://datatracker.ietf.org/doc/html/rfc6531). See my bug report: https://github.com/symfony/symfony/issues/44136

File: src/Symfony/Component/Mime/Tests/Encoder/IdnAddressEncoderTest.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Symfony\Component\Mime\Encoder;
+
+use PHPUnit\Framework\TestCase;
+
+class IdnAddressEncoderTest extends TestCase
+{
+    public function testEncodeString()
+    {
+        $this->assertEquals('test@xn--fuball-cta.test', (new IdnAddressEncoder())->encodeString('test@fußball.test'));
```
Comment (OskarStark) [inline src/Symfony/Component/Mime/Tests/Encoder/IdnAddressEncoderTest.php]: Please use assertSame 

Thanks

**Review by nicolas-grekas — APPROVED** [review]
(after @OskarStark's comment is fixed)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @j-bernard.

**Pull request merged**